src/vm/clsload.cpp

```diff
@@ -4185,7 +4210,8 @@ VOID ClassLoader::AddAvailableClassDontHaveLock(Module *pModule,
     _ASSERTE(!pModule->GetAssembly()->IsWinMD());   // WinMD files should never get into this path, otherwise provide szWinRtNamespacePrefix
 #endif
 
-    CrstHolder ch(&m_AvailableClassLock);
+    CrstHolder ch1(&m_AvailableClassCaseInsensitiveLock);
```

Comment:
We may need to switch case-insensitive table to use a different allocator to pull that off. Otherwise, it would have quadratic leak when somebody does `AddType` / `GetType case insensitive` many times...

I am wondering whether it would be easier to refactor the code so that all lower-casing is done outside the lock (we may need to retry if we decide to not lower case outside the lock, but then realize we need the lower cased key once we enter the lock).

Reply:
> I am wondering whether it would be easier to refactor the code so that all lower-casing is done outside the lock (we may need to retry if we decide to not lower case outside the lock, but then realize we need the lower cased key once we enter the lock).

I'm starting to wonder about cost/benefit of this whole thing. If we do all of this in C, yes, there will be an extra table to maintain and yes, it might be weird when dummy globalization is enabled, but that part is status quo.

This approach (calling into managed code) looks very expensive to test because we're talking about doing subtly different things for a rare profiler-related code path. I'm not even sure I got all the callsites right because doing things like constructing identifier names with macros is the norm in this codebase and findstr doesn't quite work for that. I have very little confidence in getting this right.